test/compilable/cppmangle.d

```diff
@@ -771,3 +773,165 @@ version(Windows)
       "??$test19043b@U?$test19043@$$CBD@@@@YAXU?$test19043@$$CBD@@@Z");
 }
 
+// https://issues.dlang.org/show_bug.cgi?id=16479
+//  Missing substitution while mangling C++ template parameter for functions
+version (Posix) extern (C++)
+{
+    // Make sure aliases are still resolved
+    alias Alias16479 = int;
+    Alias16479 func16479_0 (FuncT1) (FuncT1, Alias16479);
+    static assert(func16479_0!(int).mangleof == `_Z11func16479_0IiEiT_i`);
+
+    // Simple substitution on return type
+    FuncT1* func16479_1 (FuncT1) ();
+    static assert(func16479_1!(int).mangleof == `_Z11func16479_1IiEPT_v`);
+
+    // Simple substitution on parameter
+    void    func16479_2 (FuncT1) (FuncT1);
+    static assert(func16479_2!(int).mangleof == `_Z11func16479_2IiEvT_`);
+
+    // Make sure component substition is prefered over template parameter
+    FuncT1* func16479_3 (FuncT1) (FuncT1);
+    static assert(func16479_3!(int).mangleof == `_Z11func16479_3IiEPT_S0_`);
+
+    struct Array16479 (Arg) { Arg* data; }
+    struct Array16479_2 (Arg, int Size) { Arg[Size] data; }
+    struct Value16479 (int Value1, int Value2) { int data; }
+
+    // Make sure template parameter substitution happens on templated return
+    Array16479!(FuncT2) func16479_4 (FuncT1, FuncT2) (FuncT1);
+    static assert(func16479_4!(int, float).mangleof
+                  == `_Z11func16479_4IifE10Array16479IT0_ET_`);
+
+    // Make sure template parameter substitution happens with values
+    Value16479!(Value2, Value1)* func16479_5 (int Value1, int Value2) ();
+    static assert(func16479_5!(1, 1).mangleof
+                  == `_Z11func16479_5ILi1ELi1EEP10Value16479IXT0_EXT_EEv`);
+
+    // But make sure it's not substituting *too many* values
+    Value16479!(1, 1)* func16479_6 (int Value1, int Value2) ();
+    static assert(func16479_6!(1, 1).mangleof
+                  == `_Z11func16479_6ILi1ELi1EEP10Value16479ILi1ELi1EEv`);
+
+    // Or too many types
+    Array16479!(int) func16479_7 (FuncT1, FuncT2) (FuncT1);
+    static assert(func16479_7!(int, int).mangleof
+                  == `_Z11func16479_7IiiE10Array16479IiET_`);
+
+    // Also must check the parameters for template param substitution
+    void func16479_8 (FuncT1) (Array16479!(FuncT1));
+    static assert(func16479_8!(int).mangleof
+                  == `_Z11func16479_8IiEv10Array16479IT_E`);
+
+    // And non-substitution
+    void func16479_9 (FuncT1) (Array16479!(int));
+    static assert(func16479_9!(int).mangleof
+                  == `_Z11func16479_9IiEv10Array16479IiE`);
+
+    // Now let's have a bit of fun with alias parameters,
+    // starting with C functions
+    // TODO: Why is this mangled by g++:
```

Comment:
Is this working?

Reply:
Re-looking at this, it seems correct.
I'll do what Iain requested in a few hours and will update the comment.